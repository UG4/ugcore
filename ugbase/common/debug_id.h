/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef __H__UG__COMMON__DEBUG_ID_H
#define __H__UG__COMMON__DEBUG_ID_H
#include <vector>
#include <map>
#include <string>
#include "common/util/crc32.h"
#include <sstream>

namespace ug{

/// \addtogroup ugbase_common
/// \{

class DebugIDManager;



// DebugID
/**
 * A DebugID is an object to control the debug level of parts of the code.
 * It can be used in
 * - UG_SET_DEBUG_LEVEL
 * - UG_DEBUG_BEGIN/END (debugID, level)
 * - IF_DEBUG
 * - UG_DLOG(debugID, level, msg)
 *
 * ug4's standard DebugIDs are
 * APP, LIB_GRID, LIB_GRID_REFINER, LIB_DISC, LIB_DISC_ASSEMBLE, LIB_DISC_D3F,
 *		LIB_DISC_MULTIGRID, LIB_DISC_NEWTON, LIB_DISC_LINKER, LIB_DISC_TRANSFER,
 *		LIB_DISC_DISCRETE_FUNCTION, LIB_DISC_OUTPUT, LIB_DISC_OPERATOR_INVERSE,
 *		LIB_ALG_LINEAR_OPERATOR, LIB_ALG_LINEAR_SOLVER, LIB_ALG_VECTOR,
 *		LIB_ALG_MATRIX, LIB_ALG_AMG, LIB_PCL
 *
 * They are managed with \sa DebugIDManager.
 * You can define your own DebugID the following way:
 * In a cpp file:
 *
 \code{.cpp}
 DebugID DID_MYAPP("MyApp")
 DebugID DID_MYAPP_PARALLELIZATION("MyApp.Parallelization")
 \endcode{.cpp}
 *
 * Now to use it in other files you need to add there
 *
 \code{.cpp}
 extern DebugID DID_MYAPP, DID_MYAPP_PARALLELIZATION;
 ...
 UG_DLOG(DID_MYAPP, 3, "bla");
 \endcode{.cpp}
 *
 * Note that the DebugID OBJECT is called DID_MYAPP, but its string identifier
 * is "MyApp" (this is used in the script system). In LUA script you can do
 *
 * GetLogAssistant():set_debug_level("MyApp", 3)
 * or, with completion (UGIDE)
 * SetDebugLevel(debugID.MyApp, 3)
 *
 * \note The actual DebugID level is stored in DebugIDManager.
 *
 */
class DebugID
{
public:
/// registers the DebugID at DebugIDManager
	DebugID(const char *str);
	
	
	DebugID(uint32 hash) { m_hash = hash; }
/// returns the debug level via GetDebugIDManager.
	inline int get_debug_level() const;

/// sets the debug level via GetDebugIDManager.
	inline bool set_debug_level(int level);

	friend class DebugIDManager;
private:
	uint32 m_hash;
};

// the list of ug4's standard DebugIDs.
extern DebugID MAIN,
		APP,
		LIB_GRID,
		LIB_GRID_REFINER,
		LIB_DISC,
		LIB_DISC_ASSEMBLE,
		LIB_DISC_D3F,
		LIB_DISC_MULTIGRID,
		LIB_DISC_NEWTON,
		LIB_DISC_LINKER,
		LIB_DISC_TRANSFER,
		LIB_DISC_DISCRETE_FUNCTION,
		LIB_DISC_OUTPUT,
		LIB_DISC_OPERATOR_INVERSE,
		LIB_ALG_LINEAR_OPERATOR,
		LIB_ALG_LINEAR_SOLVER,
		LIB_ALG_VECTOR,
		LIB_ALG_MATRIX,
		LIB_ALG_AMG,
		LIB_PCL;


/**
 * The DebugIDManager.
 * get, set debug level, get a list of available DebugIDs etc.
 */
class DebugIDManager
{
private:
	/// returns the debug level of debugIDhash, -1 if not found
		inline int get_debug_level(uint32 debugIDhash) const
		{
			std::map<uint32, int>::const_iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return -1;
			else return (*it).second;
		}
		
		/// sets the debug level of debugIDhash if registered to 'level'
		inline bool set_debug_level(uint32 debugIDhash, int level)
		{
			std::map<uint32, int>::iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return false;
			else
			{
				(*it).second = level;
				return true;
			}
		}
public:
	///	returns a reference to the single instance of LogAssistant
		static DebugIDManager& instance();

	/// returns true if there is a DebugID with name debugIDname
		inline bool debug_id_registered(const char *debugIDname) const
		{
			return m_dbgLevels.find(crc32(debugIDname)) != m_dbgLevels.end();
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const DebugID &debugID) const
		{
			return get_debug_level(debugID.m_hash);
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const char *debugID) const
		{
			return get_debug_level(crc32(debugID));
		}
		
	/// 
		inline DebugID get_debug_id(const char *debugID) const
		{
			return DebugID(crc32(debugID));
		}
		
		inline DebugID get_or_create_debug_id(const char *debugID)
		{
			if(debug_id_registered(debugID)==false)
				register_debug_id(debugID);
			return DebugID(crc32(debugID));
		}
				

	/// sets the debug level of debugID if registered to 'level'
		inline bool set_debug_level(DebugID &debugID, int level)
		{
			return set_debug_level(debugID.m_hash, level);
		}
	/// sets the debug level of debugID if registered to 'level'
		bool set_debug_level(const char *debugID, int level);

	
	/// sets the debug level of all registered debugIDs to 'lev'
		bool set_debug_levels(int lev);

	/// returns a string describing all registered debugIDs.
		std::string get_registered_debug_IDs() const
		{
			std::stringstream str;
			str << "DebugIDs:\n";
			for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
				str << m_dbgLevelIdentifiers[i] << " : Level " << get_debug_level(m_dbgLevelIdentifiers[i].c_str()) << "\n";
			return str.str();
		}

		const std::vector<std::string> &get_registered_debug_IDs_arr() const
		{
			return m_dbgLevelIdentifiers;
		}



	bool register_debug_id(const char *debugID);

private:
	std::vector<std::string> m_dbgLevelIdentifiers;
	std::map<uint32, int> m_dbgLevels;
};

inline DebugIDManager& GetDebugIDManager()
{
	return DebugIDManager::instance();
}

inline int DebugID::get_debug_level() const
{
	return GetDebugIDManager().get_debug_level(*this);
}
inline bool DebugID::set_debug_level(int level)
{
	return GetDebugIDManager().set_debug_level(*this, level);
}

// TemporaryDebugLevel
/// a helper object to temporary set debug levels in a given scope
/**
 * A object of this class can be used to change a DebugID. The DebugID will
 * be re-set to the value before at destruction of the object.
 * This is especially useful if algorithms are calling other functions which
 * also have special DebugIDs levels. And you want to do a "debug debug".
 * example:
\code{.cpp}
  UG_SET_DEBUG_LEVEL(LIB_GRID, 1);

  // DebugID LIB_GRID is now at level 1
  {
  	  TemporaryDebugLevel tdl(LIB_GRID, 5);

  	  // DebugID LIB_GRID is now at level 5
  	  MyFunction();
  }

  // DebugID LIB_GRID is now again at 1.
\endcode
 * You can also specify differences:
 *
 \code{.cpp}
  TemporaryDebugLevel tdl(LIB_GRID, MY_DEBUG_ID, +1);
  // now LIB_GRID has the temporary debug level of MY_DEBUG_ID +1.
 */
#ifdef UG_ENABLE_DEBUG_LOGS
class TemporaryDebugLevel
{
	DebugID &did;

	int m_prevDebugLevel;
public:
	TemporaryDebugLevel(DebugID &debugID) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
	}

	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(temporaryDebugLevel);
	}

	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(debugSrc.get_debug_level()+diff);
	}

	/// resets the debugID to the previous value.
	~TemporaryDebugLevel()
	{
		did.set_debug_level(m_prevDebugLevel);
	}

	void set(int temporaryDebugLevel)
	{
		did.set_debug_level(temporaryDebugLevel);
	}
};
#else
class TemporaryDebugLevel
{

public:
	TemporaryDebugLevel(DebugID &debugID){}
	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel){}
	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0){}
	/// resets the debugID to the previous value.
	~TemporaryDebugLevel(){}
	void set(int temporaryDebugLevel){}
};
#endif

// end group ugbase_common
/// \}

}	// end namespace ug
#endif	/* __H__UG__COMMON__DEBUG_ID_H */

