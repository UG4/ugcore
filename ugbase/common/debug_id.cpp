/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#include "debug_id.h"
#include "common/log.h"
#include "common/error.h"
#include "common/assert.h"
#include <iostream>
#include <string.h>
#include "util/string_util.h"

namespace ug{

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of variables
 * of type DebugID, which are mostly global variables.
 * Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
DebugID::DebugID(const char *str)
{
	m_hash = crc32(str);
	GetDebugIDManager().register_debug_id(str);
}

DebugIDManager& DebugIDManager::instance()
{
	static DebugIDManager m;
	return m;
}


bool DebugIDManager::
set_debug_levels(int lev)
{
	for(std::map<uint32, int>::iterator it=m_dbgLevels.begin(); it != m_dbgLevels.end(); ++it)
		(*it).second = lev;
	return true;
}

bool DebugIDManager::
set_debug_level(const char *debugID, int level)
{
	int slen = strlen(debugID);
	if(slen<=0) return false;
	if(debugID[slen-1] == '*')
	{
		for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
		{
			const char *name = m_dbgLevelIdentifiers[i].c_str();
			if(WildcardMatch(name, debugID))
			{
				set_debug_level(crc32(name), level);
				//UG_LOGN(name);
			}
		}
	}
	else if(set_debug_level(crc32(debugID), level) == false)
	{
		UG_LOG("DebugID " << debugID << " not registered.\n");
		return false;
	}
	return true;
}

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of global variables
 * of type DebugID. Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
bool DebugIDManager::
register_debug_id(const char *debugID)
{
	if(debug_id_registered(debugID) == false)
	{
		m_dbgLevelIdentifiers.push_back(std::string(debugID));
		m_dbgLevels[crc32(debugID)] = -1;
		return true;
	}
	else
	{
		// not quite clear if cout is defined yet.
		// --> it should be, if we define it in this header!
		std::cout << "FATAL ERROR: DebugID "<<debugID<<" already registered." << std::endl;
		// note that this could be caused by double-registering libraries.
		UG_THROW("FATAL ERROR: DebugID "<<debugID<<" already registered.");
		return false;
	}
}

}
