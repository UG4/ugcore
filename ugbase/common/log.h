/*
 * Copyright (c) 2010-2015:  G-CSC, Goethe University Frankfurt
 * Authors: Andreas Vogel, Sebastian Reiter
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef __H__UG__COMMON__LOG__
#define __H__UG__COMMON__LOG__

#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include "util/ostream_buffer_splitter.h"
#include "types.h"
#include "ug_config.h"

#include <vector>
#include <map>
#include "common/util/crc32.h"
#include "debug_id.h"


//	in order to support VRL logs, we're including bindings_vrl.h
//  this is necessary to get access to the JVM environment
#ifdef UG_FOR_VRL
		#include <sstream>
		#include "bindings/vrl/messaging.h"
#endif

namespace ug{

/// \addtogroup ugbase_common
/// \{

	const uint64 UNIT_KILO     = 1024;                   // 2^{10}
	const uint64 UNIT_MEGA     = UNIT_KILO * 1024;       // 2^{20} =                 1'048'576
	const uint64 UNIT_GIGA     = UNIT_MEGA * 1024;       // 2^{30} =             1'073'741'824
	const uint64 UNIT_TERA     = UNIT_GIGA * 1024ll;     // 2^{40} =         1'099'511'627'776
	const uint64 UNIT_PETA     = UNIT_TERA * 1024ll;     // 2^{50} =     1'125'899'906'842'624
	const uint64 UNIT_EXA      = UNIT_PETA * 1024ll;     // 2^{60} = 1'152'921'504'606'846'976
	//const uint64_t UNIT_ZETTA    = UNIT_EXA  * 1024ll;   // 2^{70} -- too big for 64 bit 'long long int'!
	
	const uint64 UNIT_KILO_SI  = 1000;                   // 10^{ 3}
	const uint64 UNIT_MEGA_SI  = UNIT_KILO_SI * 1000;    // 10^{ 6}
	const uint64 UNIT_GIGA_SI  = UNIT_MEGA_SI * 1000;    // 10^{ 9}
	const uint64 UNIT_TERA_SI  = UNIT_GIGA_SI * 1000ll;  // 10^{12}
	const uint64 UNIT_PETA_SI  = UNIT_TERA_SI * 1000ll;  // 10^{15}
	const uint64 UNIT_EXA_SI   = UNIT_PETA_SI * 1000ll;  // 10^{18}
	//const uint64_t UNIT_ZETTA_SI = UNIT_EXA_SI  * 1000ll;// 10^{21} -- too big for 64 bit 'long long int'!



// LogAssistant
/**
 * This class provides infrastructure for logging. It separates the log messages
 * in different levels:
 *
 * logger() returns the output stream for normal User information output.
 * debug_logger() returns the output stream for debug informations.
 *
 * Debug messages are grouped by tags and debug levels. It is intended, that
 * only messages are printed, when the current level is equal or greater than
 * the debug level chosen in the code.
 *
 * Please note that this class operates on std::clog. Thus, if output-options
 * are changed (e.g. file-logging enabled) the stream buffer on which clog
 * operates will change too.
 */
class UG_API LogAssistant
{
	public:
	///	returns a reference to the single instance of LogAssistant
		static LogAssistant& instance();

	///	flushes all buffers before deletion
		~LogAssistant();

	/// enables or disables file output.
	/** a filename can be specified. Default is 'uglog.log'.
	 *	Please note that only the filename given at the first call is considered.
	 *	Filelogging is disabled by default.
	 */
		bool enable_file_output(bool bEnable, const char* filename = "uglog.log");

	/// renames output file if opened.
		bool rename_log_file(const char * newname);

	///	enables or disables terminal output.
	/**	terminal output is enbled by default.*/
		bool enable_terminal_output(bool bEnable);

	/// returns the debug output stream
		inline std::ostream& debug_logger();

	/// returns the normal output stream
		inline std::ostream& logger();

	///	returns the error output stream
	/**	Note that error-logs are not immediately visible. Instead they are
	 * gathered and can be printed in one block using LogAssistant::flush_error_log().*/
		inline std::ostream& error_logger();

	///	outputs all gathered error-logs to the standard logger and clears the error log.
		void flush_error_log();

	/// sets the debug level of all tags to 'lev'
		bool set_debug_levels(int lev)
		{
			return GetDebugIDManager().set_debug_levels(lev);
		}

	/// returns the debug level of debugID
		inline int get_debug_level(DebugID &debugID) const
		{
			return GetDebugIDManager().get_debug_level(debugID);
		}
	/// returns the debug level of debugID
		inline int get_debug_level(const char *debugID) const
		{
			return GetDebugIDManager().get_debug_level(debugID);
		}
		/// returns the debug level of debugID
		int get_debug_level_noninline(const char *debugID) const;
		
		
	/// sets the debug level of debugID
		inline bool set_debug_level(DebugID &debugID, int level)
		{
			return GetDebugIDManager().set_debug_level(debugID, level);
		}
		
	/// returns the debug level of debugID
		inline bool set_debug_level(const char* debugID, int level)
		{
			return GetDebugIDManager().set_debug_level(debugID, level);
		}
		
		bool set_debug_level_noninline(const char* debugID, int level);
		
	/// returns the debug level of debugID
		inline std::string get_registered_debug_IDs()
		{
			return GetDebugIDManager().get_registered_debug_IDs();
		}

	///	sets the output-process in a parallel environment. Default is 0.
	/**	pass a procRank of -1 to enable output on all processes.*/
		void set_output_process(int procRank);

	///	returns the rank of the current output-process or -1 if all procs perform output.
		inline int get_output_process();

	///	returns true if the current process is an output process.
	/**	This is always true, if the application is executed in a serial environment.*/
		bool is_output_process();

	///	returns the process rank of the underlying process (same as pcl::ProcRank)
		int get_process_rank();

		void flush();

	protected:
	///	updates and sets stream buffers based on current options.
	/**	Note that this method changes the buffer on which clog works.*/
		void update_ostream();

	///	opens the local logFile, if the process is the output-process
		bool open_logfile();

	private:
	/// Constructor, resets all debug levels to -1
	/**	Constructor is private since only one instance of LogAssistant may exist.*/
		LogAssistant();

	///	Prevent copying through private copy constructor.
		LogAssistant(const LogAssistant&);

	///	Performs some initialization
		void init();
	private:
	//	streams
		std::streambuf*	m_emptyBuf;
		std::streambuf*	m_logBuf;
		std::streambuf*	m_fileBuf;
		std::streambuf*	m_splitBuf;

		EmptyStreamBuffer		m_emptyBufInst;
		OStreamBufferSplitter	m_splitBufInst;

		std::string 			m_logFileName;
		std::ofstream			m_fileStream;
		std::stringstream		m_errStream;

		bool m_terminalOutputEnabled;
		bool m_fileOutputEnabled;

		int m_outputProc;

};

// returns singleton instance of LogAssistant
inline LogAssistant& GetLogAssistant();

/// returns number 'size' in a more human readable format (using IEC binary prefixes)
inline std::string ConvertNumber (uint64_t size, unsigned int width,
                                  unsigned int numDisplayedDigits);

/// returns number 'size' in a more human readable format (using SI prefixes)
inline std::string ConvertNumberSI (uint64_t size, unsigned int width,
                                    unsigned int numDisplayedDigits);

// end group ugbase_common
/// \}

} // end namespace ug


////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
// DEBUG LOG
////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////


/// \addtogroup ugbase_common
/// \{

// Usage:
/* The following macros can be used to control debug messages.
 * To use them the define 'UG_ENABLE_DEBUG_LOGS' must be set. Otherwise nothing will be done (no runtime overhead).
 *
 * UG_SET_DEBUG_LEVEL(__debugID__, level)  	- sets the debug levels of DebugID __debugID__ to level 'level'
 * UG_RESET_DEBUG_LEVELS()					- sets the debug level of all Tags to -1
 * UG_SET_DEBUG_LEVELS(level)				- sets the debug levels of all DebugIDs to level 'level'
 * UG_DLOG(__debugID__, level, msg)			- prints the message "msg" to the debug log stream, if current level of DebugID __debugID__ is >= level
 *
 * __debugID__ has to be of type DebugID.
 * For a list of standard DebugIDs and how to define your own, see debug_id.h .
 *
 *	Example:
 *
	UG_RESET_DEBUG_LEVELS();
	UG_DLOG(MAIN, 0, "DLOG on level 0.\n");		// no message printed
	UG_SET_DEBUG_LEVEL(MAIN, 0);
	UG_DLOG(MAIN, 0, "DLOG on level 0.\n");		// message printed
	UG_DLOG(MAIN, 1, "DLOG on level 1.\n");		// no message printed
 */
#ifdef UG_ENABLE_DEBUG_LOGS
	#define UG_SET_DEBUG_LEVEL(__debugID__, level)		{ug::GetDebugIDManager().set_debug_level(__debugID__, level);}
	#define UG_RESET_DEBUG_LEVELS()						{ug::GetDebugIDManager().set_debug_levels(-1);}
	#define UG_SET_DEBUG_LEVELS(level)					{ug::GetDebugIDManager().set_debug_levels(level);}
	#define UG_DEBUG_BEGIN(__debugID__, level)			{ if(ug::GetDebugIDManager().get_debug_level(__debugID__) >= level) {
	#define UG_DEBUG_END(__debugID__, level)			}; }
	#define IF_DEBUG(__debugID__, level) 				if(ug::GetDebugIDManager().get_debug_level(__debugID__) >= level)

	#define UG_DLOG(__debugID__, level, msg) \
			{if(ug::GetDebugIDManager().get_debug_level(__debugID__) >= level)\
			{ug::GetLogAssistant().debug_logger() << msg; ug::GetLogAssistant().debug_logger().flush();}}

	#define UG_DLOGN(__debugID__, level, msg)			UG_DLOG(__debugID__, level, msg << "\n");

	#define UG_DLOG_ALL_PROCS(__debugID__, level, msg)	\
			{if(ug::GetDebugIDManager().get_debug_level(__debugID__) >= level)\
			{ug::LogAssistant& la = ug::GetLogAssistant(); int op = la.get_output_process();\
			la.set_output_process(-1); la.debug_logger() << "[Proc " << la.get_process_rank() << "]: "\
			<< msg << std::flush; la.set_output_process(op);}}
#else
	#define UG_SET_DEBUG_LEVEL(__debugID__, level)		{}
	#define UG_RESET_DEBUG_LEVELS()						{}
	#define UG_SET_DEBUG_LEVELS(level)					{}
	#define UG_DLOG(__debugID__, level, msg)			{}
	#define UG_DLOGN(__debugID__, level, msg)			{}
	#define UG_DLOG_ALL_PROCS(__debugID__, level, msg)	{}
	#define UG_DEBUG_BEGIN(__debugID__, level)			{ if(1==0) {
	#define UG_DEBUG_END(__debugID__, level)			}; }
	#define IF_DEBUG(__debugID__, level)				if(1==0)
#endif

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
// WARNING LOG
////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

// Warning and debug logs are disabled when compiling for release
// defines (currently here, should be compiling options)
#ifndef NDEBUG
	#define UG_ENABLE_WARNINGS
#else /* NDEBUG */
	#undef UG_ENABLE_WARNINGS
#endif /* NDEBUG*/

// Usage:
/* The following macro can be used to print warning messages.
 * To use it the define 'UG_ENABLE_WARNINGS' must be set. Otherwise nothing will
 * be done (no runtime overhead).
 *
 * UG_WARNING(msg)  			- prints a warning to the normal output stream
 */
#ifdef UG_ENABLE_WARNINGS
	#define UG_WARNING(msg) {ug::GetLogAssistant().logger() << "UG_WARNING in "\
								<< __FILE__ << " at line " << __LINE__ << ": " \
								<< msg << std::flush;}
#else
	#define UG_WARNING(msg) {}
#endif


// Usage:
/* The following macro can be used to print warning messages if condition is met.
 * To use it the define 'UG_ENABLE_WARNINGS' must be set. Otherwise nothing will
 * be done (no runtime overhead).
 *
 * UG_WARNING(cond, msg)  		- prints a warning to the normal output stream
 */
#ifdef UG_ENABLE_WARNINGS
	#define UG_COND_WARNING(cond, msg) { if (cond) {ug::GetLogAssistant().logger() << "UG_WARNING in "\
							<< __FILE__ << " at line " << __LINE__ << ": " \
							<< msg << std::flush;} }
#else
	#define UG_COND_WARNING(cond, msg) {}
#endif





////////////////////////////////////////////////////////////////////////////////
// LOG
/**
 * UG_LOG(msg)  		- prints a message to the normal output stream
 */
#ifdef UG_FOR_VRL
	#define VRL_LOG(msg) {std::stringstream __ss; __ss << "" << msg;\
						  ug::vrl::MessageBuffer::addMessage(__ss.str());}
#else
	#define VRL_LOG(msg)
#endif

#define UG_LOG(msg) {ug::GetLogAssistant().logger() << msg << std::flush; VRL_LOG(msg);}
#define UG_COND_LOG(cond, msg) { if (cond) { UG_LOG(msg); } }
#define UG_LOGN(msg) UG_LOG(msg << "\n")
#define UG_COND_LOGN(cond, msg) { if (cond) { UG_LOGN(msg); } }
#define UG_LOG_ALL_PROCS(msg) {ug::LogAssistant& la = ug::GetLogAssistant();\
							   int op = la.get_output_process(); la.set_output_process(-1);\
							   la.logger() << "[Proc " << std::setw(3) << la.get_process_rank() << "]: "\
							   << msg << std::flush; VRL_LOG(msg); la.set_output_process(op);}

////////////////////////////////////////////////////////////////////////////////
// LOG
/**
 * UG_ERR_LOG(msg)  prints msg to LogAssistant::error_logger().
 * 					Those messages are not immediately flushed. Instead they can
 * 					be flushed to the standard logger using LogAssistant::flush_error_log
 */
#define UG_ERR_LOG(msg) {ug::GetLogAssistant().error_logger() << msg; VRL_LOG(msg);}

// include implementation
#include "log_impl.h"

// end group ugbase_common
/// \}

#endif /* __H__UG__COMMON__LOG__ */
