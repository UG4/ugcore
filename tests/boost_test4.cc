
#include "lib_algebra/cpu_algebra/sparsematrix_impl.h"
#include "lib_algebra/graph_interface/sparsematrix_boost.h"

#include "lib_algebra/graph_interface/undirected_boost.h"

// shared lib?
#include "common/log.cpp"
#include "common/debug_id.cpp"
#include "common/assert.cpp"
#include "common/util/crc32.cpp"
#include "common/util/ostream_buffer_splitter.cpp"
#include "common/util/string_util.cpp"

#include "boost/graph/graph_utility.hpp"

static const int N=7;

int main()
{
	typedef ug::SparseMatrix<double> T;
	typedef ug::UndirectedMatrix<T> UM;
	T M;

	M.resize_and_clear(N, N);

	M(0, 0) = 2.;
	M(0, 6) = -1.;
	for(unsigned i=1; i<N; ++i){
		M(i-1, i) = 1.;
	}
	M(5, 5) = 2.;
	M(4, 3) = 4.;

	/* 2  1  0  0  0  0 -1
	 * e  0  1
	 *    e  0  1
	 *       e  0  1
	 *          4  0  1
	 *             e  2  1
	 * e              e  0
	 */

	// lower fill1 0 1
	// lower fill0 0 6
	// lower fill1 1 2
	// lower fill1 2 3
	// lower fill1 4 5
	// lower fill0 5 6

	UM b(&M);

	boost::print_graph(b);

	{
		auto e = boost::vertices(b);
		for(; e.first!=e.second; ++e.first){
			auto v = *e.first;
			std::cout << v << " " << boost::out_degree(v, b) << " " << boost::in_degree(v, b) << "\n";
		}
	}

	{
		std::cout << "=== adj nodes ===\n";
		auto e = boost::adjacent_vertices(1, b);
		for(; e.first!=e.second; ++e.first){
			std::cout << *e.first << ":" <<
	//		" -- " << boost::get(wtmap, edg) <<
			"\n";
		}
	}

#if 0
	{
		std::cout << "=== out edges ===\n";
		auto e = boost::out_edges(1, b);
		for(; e.first!=e.second; ++e.first){
			boost::graph_traits<UM>::edge_descriptor const& edg = *e.first;
			std::cout << boost::source(edg, b) << ":" << boost::target(edg, b) <<
	//		" -- " << boost::get(wtmap, edg) << 
			"\n";
		}
	}
#endif
}
