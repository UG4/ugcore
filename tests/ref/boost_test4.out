0 <--> 0 1 6 
1 <--> 2 0 
2 <--> 3 1 
3 <--> 4 2 
4 <--> 3 5 
5 <--> 5 6 4 
6 <--> 0 5 
0 3 3
1 2 2
2 2 2
3 2 2
4 2 2
5 3 3
6 2 2
=== adj nodes ===
2:
0:
